Order Statuses Refactor

kloon opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
This PR refactors the single order status into new order status types, namely the main order status, payment status, fulfillment status and delivery status.

### TODO:
* [x] Create a new custom table to store new order statuses
* [x] Implement getters and setters for new statuses
* [x] Implement reading and saving to new table in the order datastore
* [x] Add deprecation notices to the existing order status functions in the datastore
* [ ] Add backward compatibility to the existing order status functions to work with the new statuses
* [ ] Write an upgrade routine to backfill new statuses for existing orders
* [ ] Implement UI changes for new order statuses and remove existing order status UI
* [ ] Update calls throughout codebase to existing order status functions to use the new order status functions
* [ ] Add unit tests for new and existing order status functions
* [ ] Update E2E tests to accommodate new order statuses


<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes # .

### How to test the changes in this Pull Request:

1.
2.
3.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> New - Introduce new order statuses to represent the overall order, payment, fulfillment, and delivery status of an order.

kloon committed:
Add order_status table structure

kloon committed:
Add wc_order_statuses to table list

kloon committed:
Define helper functions with filters for defining and retrieving payment, fulfillment and shipping statuses.

kloon committed:
Add all new payment statuses

kloon committed:
Add all fulfillment statuses

kloon committed:
Shipping status to delivery status, add all delivery statuses

kloon committed:
Update order status table to use delivery_status and not shipping_status

kloon committed:
Add new order statuses

kloon committed:
Add order helper methods for validating and transitioning all the new status types.

kloon committed:
Update order status table data structure

kloon committed:
Add new wc_get_order_statuses_from_legacy_status function decleration to working out mapping of old single order status to new status types. Hook into order update routine in DS and call a new update_order_statuses_table function which updates the new order status table. Add a Unit test for wc_get_order_statuses_from_legacy_status.

kloon committed:
Order status getters and setters, read order status data.

kloon committed:
Add missing order_status property to data

kloon committed:
Deprecate get_status and set_status in WC_Abstract_Order

kloon committed:
Update todo about adding backward compatibility to get_status and set_status instead

kloon committed:
Update some since tags, start backward compatibility for set_status

kloon committed:
Fix table create issue and missing variable when referencing table name in query

kloon committed:
Fix order_id key to be unique, change update/insert statement for order statuses.

kloon committed:
Use order object instead of passing id

kloon committed:
Add update routine methods and hook into update routine

kloon committed:
Add new statuses

kloon committed:
Add new WC_Order::needs_fulfillment function that replaces WC_Order::needs_processing and deprecate it.

kloon committed:
Add needs_filfillment check to set_fulfillment_status method to ensure we do not set virtual orders to unfulfilled

kloon committed:
Change fulfillment_status table column to allow for nulls and default to empty string since virtual orders does not have fulfillment status

kloon committed:
Add new fulfill order and complete order actions to the order actions dropdown.

kloon committed:
Update mark as complete order action text

kloon committed:
Add payment and fulfillment statuses to the order list table

kloon committed:
Remove order status dropdown from order page

kloon committed:
Add payment and fulfillment marks on order meta box next to heading

kloon committed:
Refactor WC_Order::needs_payment to look at the payment status and unpaid and awaiting-authentication payment statuses

kloon committed:
Update deprecated notice to show correct class where needs_fulfillment is located

claudiosanches committed:
Fixed filter names in wc-order-functions.php

Also makes wc_order_statuses deprecated

thenbrent commented on includes/class-wc-order.php:
```diff
@@ -1449,8 +1449,8 @@ public function get_downloadable_items() {
 	public function needs_payment() {
-		$valid_order_statuses = apply_filters( 'woocommerce_valid_order_statuses_for_payment', array( 'pending', 'failed' ), $this );
-		return apply_filters( 'woocommerce_order_needs_payment', ( $this->has_status( $valid_order_statuses ) && $this->get_total() > 0 ), $this, $valid_order_statuses );
+		$valid_payment_statuses = apply_filters( 'woocommerce_valid_order_statuses_for_payment', array( 'unpaid', 'awaiting-authentication' ), $this );
```
Could this filter be deprecated and replaced with a new filter so that `'pending'` / `'failed'` could still passed to the old filter, and `'unpaid'` / `'awaiting-authentication'` added to the new filter?

That would help maintain backward compatibility for 3rd party code attaching to `'woocommerce_valid_order_statuses_for_payment'` and checking for those old statuses.

kloon replied:
Thanks for the feedback @thenbrent, I addressed it in 75daa70 and 4f5413c

I will keep this in mind with all the other filters I touch and make the same changes accordingly.

kloon committed:
Deprecate woocommerce_valid_order_statuses_for_payment in favor of new woocommerce_valid_pending_payment_statuses filter.

kloon committed:
Fix array order for deprecation of woocommerce_valid_order_statuses_for_payment, had it the wrong way around.

kloon committed:
Merge branch 'master' into new/order-statuses-refactor

kloon committed:
Comment for woocommerce_valid_pending_payment_statuses filter

kloon committed:
Deprecate woocommerce_order_item_needs_processing in favor of woocommerce_order_item_needs_fulfillment

kloon committed:
Fix PHPCS issues, update since tags for new functions/methods, add docblocks to new filters

kloon committed:
Merge branch 'master' into new/order-statuses-refactor

kloon committed:
Introduce WC_Order::set_payment_status method

kloon committed:
Add new WC_Order::payment_status_transition method and hook it up to the save call.

kloon committed:
Add new WC_Order::fulfillment_status_transition method and hook it up the the save method for handing the fulfillment status hooks.

kloon committed:
Add comment about removing the woocommerce_order_payment_status_changed action from WC_Order::status_transition

kloon committed:
Add WC_Order::delivery_status_transition method and hook it up to the save method for handiling delivery status actions and notes

kloon committed:
Update since tags for new functions.
Override set_payment_status, set_fulfillment_status and set_delivery_status methods in WC_Order to allow for handling the status transitions.
Add new maybe_set_date_fulfilled, maybe_set_date_delivered in WC_Order for checking if we should set the dates for these statuse types.
Add new set_date_fulfilled and set_date_delivered methods in WC_Order for setting dates when those statuses reaches completed.

kloon committed:
Add new WC_Order::update_payment_status, WC_Order::update_fulfillment_status and WC_Order::update_delivery_status public methods for use to update the statuses of the orders.

kloon committed:
Update WC_Order_Data_Store_CPT::get_unpaid_orders to make use of new payment status

kloon committed:
Update all usage of 3.9.0 to 4.0.0

kloon committed:
Fix type in WC_Order::delivery_status_transition method variable

rodrigoprimo commented:
@kloon what is the status of this PR? Is this something that you still plan to work on or should we close it?

peterfabian commented:
Closing down for now, we can reopen if this project comes back.
